File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.java
import java.util.*;

class Solution {
    
    Set<String> check;
    boolean[] visited;
    int answer;
    
    public int solution(String[] user_id, String[] banned_id) {
        check = new HashSet<>();
        visited = new boolean[user_id.length];
        answer = 0;
        
        findBanList(user_id, banned_id, 0);   
        return check.size();
    }
    
    private void findBanList(String[] user_id, String[] banned_id, int n) {
        if (n == banned_id.length) {
            StringBuilder sb = new StringBuilder();
            int cnt = 0;
            for (int i = 0; i < user_id.length; i++) {
                if (visited[i]) {
                    cnt++;
                    sb.append(user_id[i]).append(", ");
                }
            }
            
            if (cnt == banned_id.length) {
                check.add(sb.toString());
            }
            return;
        }
        
        String banFormat = banned_id[n];
        for (int i = 0; i < user_id.length; i++) {
            if (!visited[i] && availableBan(banFormat, user_id[i])) {
                visited[i] = true;
                n++;
                findBanList(user_id, banned_id, n);
                n--;
                visited[i] = false;   
            }  
        }
    }
    
    private boolean availableBan(String banFormat, String user) {
        if (banFormat.length() != user.length()) {
            return false;
        }
        
        char[] banChars = banFormat.toCharArray();
        char[] userChars = user.toCharArray();
        
        for (int i = 0; i < banFormat.length(); i++) {
            char banChar = banChars[i];
            char userChar = userChars[i];
            
            if (banChar == '*') {
                continue;
            }
            
            if (banChar != userChar) {
                return false;
            }
        }
        
        return true;
    }
}

File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.py
# 8개의 user_id에서 banned_id의 길이 만큼의 조합을 구한다.
# 그조합에 banned_id에 적합한지 비교한다.
# 된다면 answer의 값을 증가시킨다.

from itertools import combinations

def solution(user_id, banned_id):
    answer = [0]
    select = len(banned_id)
    
    senarios = list(combinations(user_id, select))
    
    sequnce = [0] * select
    visited = [0] * select
    check = set()
    def isAvailable(senario, banned_id, sequnce, n):
        if n == len(sequnce):
            available = True
            for i in range(n):
                user = senario[i]
                ban = banned_id[sequnce[i]]
                if not isMatch(ban, user):
                    available = False
                    break
            if available:
                check.add(senario)
            return
        
        for i in range(len(senario)):
            if visited[i] == 0:
                visited[i] = 1
                sequnce[n] = i
                isAvailable(senario, banned_id, sequnce, n + 1)
                visited[i] = 0
    
    for senario in senarios:
        isAvailable(senario, banned_id, sequnce, 0)   
    return len(check)
            
def isMatch(ban, user):
    # 길이가 다르면 False
    if len(ban) != len(user):
        return False
    ban = list(ban)
    user = list(user)
    
    # 글자 하나씩 매칭
    for i in range(len(ban)):
        if ban[i] == "*":
            continue
        
        if ban[i] != user[i]:
            return False
    
    return True
